be/src/exprs/string_functions.cpp

```diff
@@ -1820,84 +1831,10 @@ static inline void trim_per_slice(const BinaryColumn* src, const size_t i, Bytes
-
-        const auto num_rows = src->size();
-        raw::make_room(&dst_offsets, num_rows + 1);
-        dst_offsets[0] = 0;
-        dst_bytes.reserve(dst_bytes.size());
-
-        size_t i = 0;
-        const auto sample_num = std::min(num_rows, 100ul);
-        size_t spaces_num = 0;
-
-        for (; i < sample_num; ++i) {
-            trim_per_slice<TRIM_LEFT, TRIM_SIMD_NONE, true>(src, i, &dst_bytes, &dst_offsets, &spaces_num, nullptr);
-        }
-        // when the average number of leading spaces in the sample is greater than simd_threshold,
-        // SIMD optimization is enabled.
-        if (spaces_num < simd_threshold * spaces_num) {
-            for (; i < num_rows; ++i) {
-                trim_per_slice<TRIM_LEFT, TRIM_SIMD_NONE, false>(src, i, &dst_bytes, &dst_offsets, nullptr, nullptr);
-            }
-        } else {
-            for (; i < num_rows; ++i) {
-                trim_per_slice<TRIM_LEFT, TRIM_SIMD_LEFT, false>(src, i, &dst_bytes, &dst_offsets, nullptr, nullptr);
-            }
-        }
-        return dst;
-    }
-};
-
-template <size_t simd_threshold>
-struct AdaptiveRTrimFunction {
-    template <LogicalType Type, LogicalType ResultType>
-    static inline ColumnPtr evaluate(const ColumnPtr& column) {
-        auto* src = down_cast<BinaryColumn*>(column.get());
-
-        auto dst = RunTimeColumnType<TYPE_VARCHAR>::create();
-        auto& dst_offsets = dst->get_offset();
-        auto& dst_bytes = dst->get_bytes();
-
-        const auto num_rows = src->size();
-        raw::make_room(&dst_offsets, num_rows + 1);
-        dst_offsets[0] = 0;
-        dst_bytes.reserve(dst_bytes.size());
-
-        size_t i = 0;
-        const auto sample_num = std::min(num_rows, 100ul);
-        size_t spaces_num = 0;
-
-        for (; i < sample_num; ++i) {
-            trim_per_slice<TRIM_RIGHT, TRIM_SIMD_NONE, true>(src, i, &dst_bytes, &dst_offsets, nullptr, &spaces_num);
-        }
-        // when the average number of trailing spaces in the sample is greater than simd_threshold,
-        // SIMD optimization is enabled.
-        if (spaces_num < simd_threshold * sample_num) {
-            for (; i < num_rows; ++i) {
-                trim_per_slice<TRIM_RIGHT, TRIM_SIMD_NONE, false>(src, i, &dst_bytes, &dst_offsets, nullptr, nullptr);
-            }
-        } else {
-            for (; i < num_rows; ++i) {
-                trim_per_slice<TRIM_RIGHT, TRIM_SIMD_RIGHT, false>(src, i, &dst_bytes, &dst_offsets, nullptr, nullptr);
-            }
-        }
-        return dst;
-    }
-};
-
-template <size_t simd_threshold>
+template <TrimType trim_type, size_t simd_threshold, bool trim_single>
 struct AdaptiveTrimFunction {
-    template <LogicalType Type, LogicalType ResultType>
-    static inline ColumnPtr evaluate(const ColumnPtr& column) {
+    template <LogicalType Type, LogicalType ResultType, typename Args>
+    static ColumnPtr evaluate(const ColumnPtr& column, Args remove) {
```

Comment:
Maybe use `Args&&` or `const Args&` to avoid copying `remove`?

Reply:
good advice